Support SSL connections in FrontendStreams

johto opened the pull request:
I figured this was a nice way to support SSL in the frontend piece.  The idea is to allow the user of FrontendStream to "upgrade" the connection to an SSL one after receiving SSLRequest.  A convenience method AcceptSSL is provided, but in a real application the user might want to wrap the net.Conn in a bufio writer in the ConnectionUpgradeFunc as well, so the low-level interface is exported.

Thoughts?  Better ideas?

johto commented:
Oh, I forgot: [here](https://gist.github.com/johto/9c04497cccfffb6a7529) is the snippet I used to test this approach.

uhoh-itsmaciek commented on core/stream.go:
```diff
@@ -232,6 +236,38 @@ func (c *MessageStream) Flush() error {
+	return connState, err
+}
+
+func (c *MessageStream) ConnectionUpgrade(up ConnectionUpgradeFunc) (err error) {
```
What do you think about renaming this function to `UpgradeConnection` (I think the `ConnectionUpgradeFunc` type name works as is)?

johto replied:
That sounds like a good idea.  Will amend.

uhoh-itsmaciek commented:
Nice, I think this works. Just a minor naming comment inline.

johto commented:
I now also noticed that there's a semicolon on line 47.  Will have to remove that one as well.

johto commented:
Updated the code.  How does this look?

Also, any thoughts what to do with example/simpleproxy.go?  Should it just generate some random certs at the beginning and support SSL?  Or just reject them cleanly?

cbandy commented on core/stream.go:
```diff
@@ -232,6 +236,39 @@ func (c *MessageStream) Flush() error {
+}
+
+func (c *MessageStream) UpgradeConnection(up ConnectionUpgradeFunc) error {
+	var err error
```
I'd prefer to name `err` in the result parameters in this case.

johto replied:
I had it like that, but I broke the code once by removing the "unnecessary" name.  So I figured this is more obvious to the reader.  Perhaps that's just my inexperience with Go, though.

uhoh-itsmaciek replied:
For what it's worth, I don't feel strongly one way or the other about minor style issues like that.

cbandy commented on core/stream.go:
```diff
@@ -232,6 +236,39 @@ func (c *MessageStream) Flush() error {
+		uc := tls.Server(nc, tlsConfig)
+		err := uc.Handshake()
+		if err != nil {
+			return nil, err
```
(I'm not familiar with TLS or what happens during handshake.) Is it correct/acceptable to return `nil` here? It appears this value is unconditionally assigned to `c.rw` in `UpgradeConnection()`.

johto replied:
Hmmh.  I don't think it's not correct since the connection is in an undefined state and thus unusable for any communications.  But I don't see a reason to replace it with nil either.  Changed.

cbandy commented:
Without knowing all the nomenclature for TLS nor FEBE, I would describe this feature more like "replace connection" or "replace io". The returned `ReadWriteCloser` may have nothing to do with the existing one.

> in a real application the user might want to wrap the net.Conn

Maybe "wrap" communicates the intent more clearly.

If it makes anything easier, we don't need to name the `func` type in the parameter list:

```go
func (c *MessageStream) UpgradeConnection(f func (io.ReadWriteCloser) (io.ReadWriteCloser, error)) error
```

uhoh-itsmaciek commented:
@cbandy mechanically, I think you're right--this is pretty general. Semantically, though I think this is very close to [HTTP connection upgrades](http://en.wikipedia.org/wiki/HTTP/1.1_Upgrade_header) and I hesitate to try to support something far outside of that workflow. Any thoughts, @johto?

uhoh-itsmaciek commented:
@johto looks good. Re:

> Also, any thoughts what to do with example/simpleproxy.go? Should it just generate some random certs at the beginning and support SSL? Or just reject them cleanly?

If you could generate and check in a snake-oil cert, that'd be great. Otherwise, rejecting is fine and we can expand on the example later.

johto commented:
>> Without knowing all the nomenclature for TLS nor FEBE, I would describe this feature more like "replace connection" or "replace io". The returned ReadWriteCloser may have nothing to do with the existing one.

> @cbandy mechanically, I think you're right--this is pretty general. Semantically, though I think this is very close to HTTP connection upgrades and I hesitate to try to support something far outside of that workflow. Any thoughts, @johto?

Yeah, the naming comes from HTTP connection upgrades, and while technically you could use this method to replace the connection with something completely unrelated, doing that doesn't seem like a good idea.

>> in a real application the user might want to wrap the net.Conn

> Maybe "wrap" communicates the intent more clearly.

I don't find that very descriptive for the TLS use case to be honest.  In my mind, TLS is much more than just a wrapper around the connection.

I do have to admit that UpgradeConnection isn't the perfect name, but so far I haven't come across anything I would find better, either.

johto commented:
> If it makes anything easier, we don't need to name the func type in the parameter list:

I found that that made the function signature awfully long, and even cost some readability.

cbandy commented:
> this is very close to HTTP connection upgrades and I hesitate to try to support something far outside of that workflow.
 -
> the naming comes from HTTP connection upgrades, and while technically you could use this method to replace the connection with something completely unrelated, doing that doesn't seem like a good idea.

It sounds like `AcceptSSL` is all we want right now.

> in a real application the user might want to wrap the net.Conn in a bufio writer

@johto, can you expand on this? I think the user can (should?) do that before instantiating the stream object.

johto commented:
>> in a real application the user might want to wrap the net.Conn in a buffo writer

> @johto, can you expand on this? I think the user can (should?) do that before instantiating the stream object.

A bufio writer/reader doesn't implement net.Conn, so you can't use it with crypto/tls.  You'd have to first wrap it for the handshake, then pull out the underlying net.Conn, and then re-wrap the result.  That's why I'd just prefer doing unbuffered I/O during the handshake, and wrap the Conn returned by crypto/tls.  I guess you could implement your own net.Conn on top of a buffered reader/writer, but I'm not sure why anyone would bother.

Though I'm not 100% how much one can gain by wrapping a TLS net.Conn with bufio in the first place.  But I still don't think we should give up the flexibility when there's no reason to.

cbandy commented:
> A bufio writer/reader doesn't implement net.Conn

OT: it looks like only `RemoteAddr` of `net.Conn` is being used in `crypto/tls`, and only as a *fallback* cache key on the client-side when the user fails to populate `ServerName`... :unamused: 

```
% ag 'LocalAddr|RemoteAddr|SetDeadline|SetReadDeadline|SetWriteDeadline'
tls/conn.go
65:// LocalAddr returns the local network address.
66:func (c *Conn) LocalAddr() net.Addr {
67:     return c.conn.LocalAddr()
70:// RemoteAddr returns the remote network address.
71:func (c *Conn) RemoteAddr() net.Addr {
72:     return c.conn.RemoteAddr()
75:// SetDeadline sets the read and write deadlines associated with the connection.
78:func (c *Conn) SetDeadline(t time.Time) error {
79:     return c.conn.SetDeadline(t)
82:// SetReadDeadline sets the read deadline on the underlying connection.
84:func (c *Conn) SetReadDeadline(t time.Time) error {
85:     return c.conn.SetReadDeadline(t)
88:// SetWriteDeadline sets the write deadline on the underlying connection.
91:func (c *Conn) SetWriteDeadline(t time.Time) error {
92:     return c.conn.SetWriteDeadline(t)
885:// after a fixed time limit; see SetDeadline and SetReadDeadline.

tls/handshake_client.go
106:            cacheKey = clientSessionCacheKey(c.conn.RemoteAddr(), c.config)
```

cbandy commented:
> I still don't think we should give up the flexibility when there's no reason to

And I don't think we should export features without a tangible use case.

cbandy commented:
> I still don't think we should give up the flexibility when there's no reason to

And I don't think we should export features without a tangible use case.

All that said, the code here looks good to me!

johto commented:
>> I still don't think we should give up the flexibility when there's no reason to

> And I don't think we should export features without a tangible use case.

That's not an unreasonable opinion, but I've been on the worse side of these things *way* too many times.  Even if we can't come up with a use case for something doesn't mean that there isn't one.  It's really frustrating to learn that you can't do your job properly and instead have to look for a workaround only because someone artificially prevented you from doing something because they couldn't come up with a use case.  If that means that nobody ever calls UpgradeConnection directly, I think that's acceptable, since the alternative might be that someone wants to call UpgradeConnection but can't.

I was not going to send this because I'm probably overreacting in the case of UpgradeConnection, but I'm leaving this here anyway, in hopes that it will make a difference 150 years from now.

Thanks for reviewing, much appreciated.

johto commented:
>> A bufio writer/reader doesn't implement net.Conn

> OT: it looks like only RemoteAddr of net.Conn is being used in crypto/tls, and only as a fallback cache key on the client-side when the user fails to populate ServerName...

Huh.  I guess the idea is just to provide something which is interchangeable with a TCP/UNIX/whatever connection.  I always expected it to do something more with the connection.

uhoh-itsmaciek commented:
While generally I tend to err on the side of less-is-more with respect to interface design, given that the library is already basically about accommodating unusual use cases, the more general hook does make sense, I think.

Also:

> I guess you could implement your own net.Conn on top of a buffered reader/writer, but I'm not sure why anyone would bother.

Should we maybe do this *for* the user? I guess we wouldn't even need to do buffering: just wrap the `ReadWriteCloser` in something that makes it look like a `net.Conn`:

```golang
type FakeConn struct { io.ReadWriteCloser }
func (f *FakeConn) LocalAddr() net.Addr { return nil }
func (f *FakeConn) RemoteAddr() net.Addr { return nil }
func (f *FakeConn) SetDeadline(t time.Time) error { return nil }
func (f *FakeConn) SetReadDeadline(t time.Time) error { return nil }
func (f *FakeConn) SetWriteDeadline(t time.Time) error { return nil }
```

Or maybe even return real dummy `Addr`s if `crypto/tls` is picky. This isn't pretty, but it's prettier than panicking if someone tries to AcceptSSL on a MessageStream using a non-`net.Conn` library.

Of course, we'd only do this if the `io.ReadWriteCloser` is not already a `net.Conn`.

johto commented:
> Should we maybe do this for the user? I guess we wouldn't even need to do buffering: just wrap the ReadWriteCloser in something that makes it look like a net.Conn:

Hmm.  That's not necessarily a bad idea.  Some inspiration [here](http://golang.org/src/net/pipe.go?s=462:486#L8), for example.

If we have an agreement on this, I could amend the PR among these lines.

uhoh-itsmaciek commented:
I think we can consider the lack of objections agreement...?

johto commented:
I think so.  I've got a lot on my plate right now, but I'll try and get this done during the weekend.

johto committed:
Support SSL connections in FrontendStreams

Add a new method UpgradeConnection to MessageStream.  This allows the
caller to "upgrade" the connection to an SSL one after receiving an
SSLRequest from the frontend.  Additionally provide a convenience method
which completes the entire procedure on behalf of the user.

johto commented:
How's this version?

cbandy commented:
lgtm

uhoh-itsmaciek commented:
Sorry, totally dropped the ball here. I can merge this if this is still useful.
